### PR TITLE
Add pragma(printf) to most error/warning/deprecation functions...

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
..because those functions forward to printf and friends. The remaining functions require `pragma(printf, <format index>)` because the format string is preceded by another `char*` argument.

Also add `-ignore` when building dmd if the host compiler is older than 2.092.0.

CC @WalterBright
This is blocked because  `pragma(printf)` requires `extern(C|C++)` linkage and hence forced me to make some functions `extern(C++)` for now - which is not a viable solution.
I think we should remove this restriction because `pragma(printf)` can validate the **arguments** passed to `printf` wrappers regardless of their linkage.

EDIT:
Also blocked because `ignore` does not work for `pragma(printf)`...

### CODE REVIEW

**Comment by WalterBright** [comment]
printf-style variadics should be C/C++, because if they are D style they add hidden arguments. I don't know why making them extern C++ is not viable for DMD.

I have not added this to the DMD source because, as you found, it needs to be compilable with older DMDs.

**Comment by MoonlightSentinel** [comment]
> printf-style variadics should be C/C++, because if they are D style they add hidden arguments

That may be true but the purpose of  `pragma(printf)` is to ensure that the format string and arguments match _at the call site_. This limitation does not bring any value in this regard and just limits its potential.

> I don't know why making them extern C++ is not viable for DMD.

Let me rephrase that as _not  a viable solution for everyone who wants to use `pragma(printf)`_. 


> I have not added this to the DMD source because, as you found, it needs to be compilable with older DMDs.

Hopefully that will change with the next update to the minimal host compiler (if it includes the fix for `-ignore`)

**Comment by WalterBright** [comment]
> That may be true but the purpose of pragma(printf) is to ensure that the format string and arguments match at the call site. This limitation does not bring any value in this regard and just limits its potential.

If you're using D functions, then you should be using writef, not printf.

> Let me rephrase that as not a viable solution for everyone who wants to use pragma(printf).

Why not? Adding two extra arguments, that are unused, for the D version, and then never using them doesn't make much sense.

**Comment by MoonlightSentinel** [comment]
> If you're using D functions, then you should be using writef, not printf.

Not everybody wants to use phobos (e.g. dmd itself)

> Why not? Adding two extra arguments, that are unused, for the D version, and then never using them doesn't make much sense.


That won’t always be the case and misses the point.
If a user declares „my function obeys the printf standard“ (e.g. by doing some pre-/postprocessing and forwarding to printf) he should be allowed to rely on the already implemented checks.
Requiring C(++) linkage is an arbitrary restriction and doesn‘t sit well with the idea of orthgonal features.

**Comment by WalterBright** [comment]
Just declare the D printf functions as extern (C++). There's simply no reason to use extern(D) with them.
